### ISSUE TITLE
Need example for nnets with pitch features

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1011]
We need an example scripts somewhere for recent nnet recipes (nnet2 and nnet3) with pitch features, that are suitable for online decoding.  E.g. in hkust or gale_mandarin.  I may have to do this myself... I'm not sure if there is an example anywhere.

**Comment by vijayaditya** [comment]
The nnet3 babel recipe uses pitch features, but I don't know if these are extracted online or offline. @pegahgh could you clarify ?

**Comment by pegahgh** [comment]
They are extracted offline!

**Comment by YanqiangLei** [comment]
Maybe an example script with pitch and online-cmvn process is also strongly required.

**Comment by naxingyu** [comment]
hkust would be a good example. I'm working on it.

**Comment by naxingyu** [comment]
@vijayaditya <https://github.com/vijayaditya> Which script are you 
talking about? Is it committed or in CLSP cluster?

X.


在 2016/8/30 5:48, Vijayaditya Peddinti 写道:

**Comment by vijayaditya** [comment]
This is in the Babel branch @jtrmal or @pegahgh would know more about this
script.

--Vijay

**Comment by jtrmal** [comment]
we do use pitch in babel recipes, but everything is offline.
I'd suggest going for HKUST first
I have a fresh system trained here:
/export/b18/jtrmal/LORELEI/systems/hkust-2
which can be used as a baseline + there are some improvements in the
scoring pipeline.
y.

**Comment by naxingyu** [comment]
Thank you Yenda. I do plan to work on hkust, but I want to checkout the 
way you use pitch in babel.

X.

在 2016/11/19 0:41, jtrmal 写道:
> we do use pitch in babel recipes, but everything is offline.
> I'd suggest going for HKUST first
> I have a fresh system trained here:
> /export/b18/jtrmal/LORELEI/systems/hkust-2
> which can be used as a baseline + there are some improvements in the
> scoring pipeline.
> y.
>
>
> On Fri, Nov 18, 2016 at 11:28 AM, Vijayaditya Peddinti <

**Comment by jtrmal** [comment]
I think the only magic we do is to run steps/make_plp_pitch.sh in place of
steps/make_plp.sh :)
Also, unless I'm mistaken, we don't use pitch in nnets in babel.
There is an trained turkish system including the outputs
in /export/a15/jtrmal/babel/105-turkish-flp
if you want to have a look.
y.

**Comment by danpovey** [comment]
OK, but the thing I had asked Xingyu to do involves using pitch in the
neural nets.  The trick is, you don't train the iVector extractor with the
features that include pitch, but you train the neural net on the slightly
longer features including pitch.  I'm sure there were example scripts for
this at some point, not sure where they are now.  But there is a new
mechanism now for extracting the MFCCs from the features.

Please view this email as text in case things in angle brackets do not show
up.

Xingyu, there is actually an efficient way to do this without re-dumping
features, but it involves writing a new script.  You will see that there is
a script steps/select_feats.sh.  That script should now be deprecated
[change the message and the comments] and you can write a new script called
utils/data/limit_feature_dim.sh, which uses the 'matrix-range' mechanism
that has been introduced recently.  This will be like a much simplified
version of copy_data_dir.sh with the options removed, and will have usage
like:

utils/data/limit_feature_dim.sh <feature-dim-range> <srcdir> <destdir>
e.g.: utils/data/limit_feature_dim.sh 0:40 data/train_hires_pitch
data/train_hires

The script will append the data-range in square brackets and after a comma
to the entries in the feats.scp file (e.g. [,0:40]), and then call
utils/data/normalize_data_range.pl, like subsegment_data_dir.sh does.
 [this is for generality, in case the input feats.scp already has some kind
of data range; it normalizes it so you don't have two square-bracket
expressions one after the other].
Unlike copy_data_dir.sh you don't have to do all the filtering stuff, just
copy the files, but don't copy cmvn.scp because it will need to be
recomputed.  The script will be quite short and simple.

The options to prepare_online_decoding.sh or whatever it's called will
change, you have to tell it that you're using pitch features and give it
the online-pitch-extraction config file.

Dan

**Comment by jtrmal** [comment]
yes, that's what I was trying to explain (that the babel scripts won't be of much use).
I think we were asking @pegahgh  for some experiments in that direction when she was preparing the paper, not sure if she got to it and if yes, if she still has the scripts.

**Comment by naxingyu** [comment]
I see what you mean. Thanks.

X.

**Comment by naxingyu** [comment]
There are make_mfcc_pitch{,_online}.sh scripts. I see the difference is that there is no post processing of pitch in the online script. So should I use the online script to extract feats?

**Comment by danpovey** [comment]
Yes, use the online script... the postprocessing is done in the same binary
as the pitch is generated in.
Dan

**Comment by naxingyu** [comment]
Is there anything that should be done to resolve this issue?

**Comment by danpovey** [comment]
Seems to be done-- closing.
